File: agent.ts
import { StateGraph, START, END, Annotation } from "@langchain/langgraph";
import {
  ticketIntake,
  classify,
  routeDecision,
  autoResolve,
  assignToTeam,
  notifyUser,
  updateTicket,
  fallback,
} from "./nodes.js";

// LangGraph needs the state shape declared as "Annotations" so it knows
// how to merge updates returned by each node. Since every one of our
// node functions already returns a FULL, merged state object (using
// `...state` spreads), we don't need custom merge logic — each field
// just gets overwritten with whatever the node returns.
const GraphState = Annotation.Root({
  ticket_id: Annotation<string>(),
  subject: Annotation<string>(),
  description: Annotation<string>(),
  user_email: Annotation<string>(),
  user_name: Annotation<string>(),
  category: Annotation<string>(),
  urgency: Annotation<string>(),
  confidence: Annotation<number>(),
  reasoning: Annotation<string>(),
  assignment_group: Annotation<string>(),
  resolution_status: Annotation<string>(),
  resolution_notes: Annotation<string>(),
  email_body: Annotation<string>(),
  actions_log: Annotation<string[]>(),
  error: Annotation<string>(),
});

// Build the graph: nodes first, then edges (including the two
// conditional branch points from the assignment spec).
const workflow = new StateGraph(GraphState)
  .addNode("ticket_intake", (state) => ticketIntake(state))
  .addNode("classify", classify)
  .addNode("auto_resolve", autoResolve)
  .addNode("assign_to_team", assignToTeam)
  .addNode("notify_user", notifyUser)
  .addNode("update_ticket", updateTicket)
  .addNode("fallback", fallback)

  .addEdge(START, "ticket_intake")

  // Branch 1: did intake validation fail (missing required fields)?
  .addConditionalEdges(
    "ticket_intake",
    (state) => (state.error ? "fallback" : "classify"),
    { fallback: "fallback", classify: "classify" }
  )

  // Branch 2: based on classification result, decide the path.
  .addConditionalEdges("classify", routeDecision, {
    auto_resolve: "auto_resolve",
    assign_to_team: "assign_to_team",
    fallback: "fallback",
  })

  .addEdge("auto_resolve", "notify_user")
  .addEdge("assign_to_team", "notify_user")
  .addEdge("notify_user", "update_ticket")
  .addEdge("update_ticket", END)
  .addEdge("fallback", END);

export const agent = workflow.compile();

File: agent.py
"""Assembles the LangGraph triage agent.

Flow:

    START
      -> ticket_intake
           -> (valid)   classify
           -> (invalid) fallback
      -> classify
           -> (Password Reset & high conf) auto_resolve
           -> (low confidence / failed)    fallback
           -> (otherwise)                  assign_to_team
      -> auto_resolve / assign_to_team -> notify_user -> update_ticket -> END
      -> fallback -> END
"""

from langgraph.graph import StateGraph, START, END

import nodes
from state import TicketState


def build_agent():
    """Build and compile the triage graph into a runnable agent."""
    graph = StateGraph(TicketState)

    # --- Register nodes ---
    graph.add_node("ticket_intake", nodes.ticket_intake_node)
    graph.add_node("classify", nodes.classify_node)
    graph.add_node("auto_resolve", nodes.auto_resolve_node)
    graph.add_node("assign_to_team", nodes.assign_to_team_node)
    graph.add_node("notify_user", nodes.notify_user_node)
    graph.add_node("update_ticket", nodes.update_ticket_node)
    graph.add_node("fallback", nodes.fallback_node)

    # --- Edges ---
    graph.add_edge(START, "ticket_intake")

    # Intake validates; bad input skips classification entirely.
    graph.add_conditional_edges(
        "ticket_intake",
        nodes.route_after_intake,
        {"classify": "classify", "fallback": "fallback"},
    )

    # Core routing decision after classification.
    graph.add_conditional_edges(
        "classify",
        nodes.route_decision,
        {
            "auto_resolve": "auto_resolve",
            "assign_to_team": "assign_to_team",
            "fallback": "fallback",
        },
    )

    # Both action paths converge on notification, then the final update.
    graph.add_edge("auto_resolve", "notify_user")
    graph.add_edge("assign_to_team", "notify_user")
    graph.add_edge("notify_user", "update_ticket")
    graph.add_edge("update_ticket", END)

    # Fallback terminates without notifying (input may lack a usable email).
    graph.add_edge("fallback", END)

    return graph.compile()


# Build once at import so callers can `from agent import agent`.
agent = build_agent()
